### ISSUE TITLE
JSON schema doesn't have link objects validations for Eiffel events 

### ISSUE THREAD

**Issue by Umadevi-Kapu** [eiffel-community/eiffel#148]
From https://github.com/Ericsson/eiffel/blob/master/eiffel-syntax-and-usage/the-links-object.md,
we notice that few links are specific to Few Eiffel Events, 

For Example: EiffelActivityCanceledEvent
Required : ACTIVITY_EXECUTION
Optional : CAUSE/CONTEXT, FLOW_CONTEXT

Today in schema there is no validation on the type field in links array object. We have tried to cover half of this (Few Links specific to Few events) in this pull request https://github.com/Ericsson/eiffel/pull/147.
We did not find ways in which we can cover mandatory and optional links without hardcoding the links in each event instead of type and value.

**Comment by d-stahl-ericsson** [comment]
Hi! Sorry for the late reply - I've been on vacation.

I have given this issue some thought myself. Schema validation of links would be great, but it's easier said than done. As you indicate, there are two types of checks that can be made:
1) Ensure that required links are present
2) Ensure that no links other than defined ones are present

The first would be very valuable, but I'm uncertain whether it can be done with JSON schema. Perhaps using oneOf? If I get the time I'll try it out...

The second is easily done, but unfortunately it shuts down the protocol. Additional link types should not be prohibited. If I wish to extend the protocol by defining my own DanielsCandyFlossCreatedEvent and want to reference it from other events using a CANDY_FLOSS link type, then that's perfectly fine. The current PR to this issue disallows that, though.

**Comment by d-stahl-ericsson** [comment]
My mistake. The keyword I was thinking of was "contains", not "oneOf". I'll play around a bit and see if I can make something work.

**Comment by d-stahl-ericsson** [comment]
Unfortunately the current version of jsonschema doesn't support draft 6 (where "contains" is introduced). See https://github.com/Julian/jsonschema/issues/337

I tried working around this with the following:
```
      "not": {
        "items": {
          "properties": {
            "type": {
              "not": {
                "enum": ["IUT"]
              }
            }
          }
        }
      }

Problem is I can't make that work for multiple required types (consider EiffelArtifactReusedEvent, which requires REUSED_ARTIFACT and COMPOSITION). Perhaps there's a way that I just don't see? Otherwise, I suggest we put this on hold and revisit it when jsonschema support draft 6.

What we easily can do, however, is to support validation of UUID syntax in links. I think I'll make a PR to that effect...
```

**Comment by d-stahl-ericsson** [comment]
This is related to #136, by the way.

**Comment by d-stahl-ericsson** [comment]
Quick update: still no draft6 support :( Would be neat if we could contribute to jsonschema to help fix that, but the time can be found right now.

**Comment by magnusbaeck** [comment]
AFAICT draft 6 support is available now so it should be possible to start working on this issue.

**Comment by xumakap** [comment]
We will look into it.

**Comment by e-backmark-ericsson** [comment]
This issue is now considered solved thanks to #380 . That PR only covers the aspect of validating the existence of mandatory link types. The schema does not validate that all provided link types are valid.
